client/PRUserial485/PRUserial485.py

```diff
@@ -161,101 +135,157 @@ def socket_communicate(conn_port, data_queue):
     notification_event.wait()
     notification_event.clear()
-    command, payload_recv = answers.pop(notification_event, None)
+    command, payload_recv = COMM_RESPONSE.pop(notification_event, None)
     del notification_event
 
     return command, payload_recv
 
 
+# --- PRUserial485 API ---
+
+
+def PRUserial485_set_bbb_ip_address(ip_address):
+    """Define beaglebone IP address."""
+    global BBB_IP
+    BBB_IP = ip_address
+
+
 def PRUserial485_open(baudrate=6, mode=b'M'):
     """Procedimento de inicialização da PRU."""
     if (mode in _c.AVAILABLE_MODES) and \
-        (baudrate in _c.AVAILABLE_BAUDRATES):
+       (baudrate in _c.AVAILABLE_BAUDRATES):
 
         payload = _c.COMMAND_PRUserial485_open + \
-        mode + struct.pack(">I", baudrate)
+            mode + struct.pack(">I", baudrate)
 
-        command, payload_recv = send_communication_data(payload)
+        command, payload_recv = _send_communication_data(payload)
 
-        if command == ord(_c.COMMAND_PRUserial485_open) and len(payload_recv) == 1:
-            return(ord(payload_recv))
+        if command == ord(_c.COMMAND_PRUserial485_open) and \
+           len(payload_recv) == 1:
+            return ord(payload_recv)
         else:
             return None
 
 
-def PRUserial485_address():
-    """Retorna endereco fisico da placa."""
-    # Payload: none
-    payload = _c.COMMAND_PRUserial485_address
-    command, payload_recv = send_communication_data(payload)
-    if command == ord(_c.COMMAND_PRUserial485_address) and len(payload_recv) == 1:
-        return(ord(payload_recv))
-    else:
-        return None
-
 def PRUserial485_close():
     """Encerra a PRU."""
     # Payload: none
     payload = _c.COMMAND_PRUserial485_close
-    command, payload_recv = send_communication_data(payload)
-    if command == ord(_c.COMMAND_PRUserial485_close):
-        return
+    _send_communication_data(payload)
+
+
+def PRUserial485_threads_start():
+    """."""
+    # create threads
+    thread_cmd_general = \
+        Thread(target=_socket_communicate,
+               args=(SERVER_PORT_GENERAL, QUEUE_GENERAL))
+    thread_cmd_general.setDaemon(True)
+    thread_cmd_rw = \
+        Thread(target=_socket_communicate,
+               args=(SERVER_PORT_RW, QUEUE_RW))
+    thread_cmd_rw.setDaemon(True)
+
+    # start threads
+    thread_cmd_general.start()
+    thread_cmd_rw.start()
+
+
+def PRUserial485_read():
+    """Recebe dados através da interface serial."""
+    # Payload: none
+    payload = _c.COMMAND_PRUserial485_read
+    command, payload_recv = _send_communication_data(payload)
+    data = [chr(i) for i in payload_recv]
+    if command == ord(_c.COMMAND_PRUserial485_read):
+        return data
     else:
         return None
 
 
-def PRUserial485_write(data=[], timeout=0):
+def PRUserial485_write(data=None, timeout=0):
     """Envia dados através da interface serial."""
     # Payload: TIMEOUT (4 bytes) + DATA (len(DATA) bytes)
+    if data is None:
+        data = []
```

Comment:
Lines 239/240 not needed if defining function with "def PRUserial485_write(data=[], timeout=0)"

Reply:
It is considered a good practice not using mutables as default function arguments. Visual Studio code (code editor we use) keeps warning about it. That's why I changed it.

https://stackoverflow.com/questions/366422/what-is-the-pythonic-way-to-avoid-default-parameters-that-are-empty-lists

https://docs.python.org/3/reference/compound_stmts.html#function-definitions